raiden_contracts/contracts/TokenNetwork.sol

```diff
@@ -275,296 +278,299 @@ contract TokenNetwork is Utils {
+    /// `participant1`, contained in locked transfers that will be retrieved by calling `unlock`
+    /// after the channel is settled.
+    /// @param participant2_locksroot The latest known merkle root of the pending hash-time locks
+    /// of `participant2`, used to validate the unlocked proofs.
     function settleChannel(
         uint256 channel_identifier,
         address participant1,
-        address participant2)
-        isClosed(channel_identifier)
-        timeoutOver(channel_identifier)
+        uint256 participant1_transferred_amount,
+        uint256 participant1_locked_amount,
+        bytes32 participant1_locksroot,
+        address participant2,
+        uint256 participant2_transferred_amount,
+        uint256 participant2_locked_amount,
+        bytes32 participant2_locksroot)
         public
     {
-        uint256 participant1_amount;
-        uint256 participant2_amount;
-        uint256 total_deposit;
-
         Channel storage channel = channels[channel_identifier];
 
+        // Channel must be closed
+        require(channel.state == 2);
+
+        // Settlement window must be over
+        require(channel.settle_block_number < block.number);
+
         Participant storage participant1_state = channel.participants[participant1];
         Participant storage participant2_state = channel.participants[participant2];
 
-        // Make sure the addresses are channel participant addresses
+        // Addresses must be channel participants
         require(participant1_state.initialized);
         require(participant2_state.initialized);
 
+        require(verifyBalanceProofData(
+            channel_identifier,
+            participant1,
+            participant2,
+            participant1_transferred_amount,
+            participant1_locked_amount,
+            participant1_locksroot
+        ));
+
+        require(verifyBalanceProofData(
+            channel_identifier,
+            participant2,
+            participant1,
+            participant2_transferred_amount,
+            participant2_locked_amount,
+            participant2_locksroot
+        ));
+
+        // We cannot use anymore local variables here because we get a "Stack too deep" error
+        // `participant2_transferred_amount` is the amount of tokens that `participant1`
+        // needs to receive. `participant1_transferred_amount` is the amount of tokens that
+        // `participant2` needs to receive
+        (
+            participant2_transferred_amount,
+            participant1_transferred_amount
+        ) = getSettleTransferAmounts(
+            uint256(participant1_state.deposit),
+            uint256(participant1_transferred_amount),
+            uint256(participant1_locked_amount),
+            uint256(participant2_state.deposit),
+            uint256(participant2_transferred_amount),
+            uint256(participant2_locked_amount)
+        );
+
+        // Remove the channel data from storage
```

Comment:
Arguments for my approach:

- people may choose to never `unlock` if the value is way less than transaction cost
- I wanted to make `settleChannel` cost as less gas as possible - because this **will** be used more frequently than `unlock`
- I see no reason to stop people creating a new channel after settling, even if there are unlocked tokens in the contract. These tokens are linked to a unique `channel_identifier` inside the contract and just need to be stored somewhere in the `raiden` client - to keep track of them. They are no longer functioning channels.

Reply:
I agree with all except this:

> I wanted to make settleChannel cost as less gas as possible - because this will be used more frequently than unlock

Moving the `delete` to a different point will have the "savings" too.

Anyways, `delete`ing here or afterwards is not really what wories me. As I understanding, the channel *still exists* after settle, specially because we still have operations to do, namely unlock.

It's just weird to have the channel "gone" even though we may still execute some operations on it. With that said, the core of my point is: It would be good to have information about the channel lifecycle until it has ended.

With that said, it's fine with me if the `delete` stays here, I just don't know how to introduce information about the lifecycle of the channel using the available structures (we may do it in a different way, but it would be nice to define it, perhaps using the events?)

Reply:
Started a lifecycle issue here: https://github.com/raiden-network/raiden-contracts/issues/48